File: ts/src/middleware.ts
/**
 * Middleware — provider-agnostic compression orchestrator.
 *
 * The Middleware is what the SDK wrappers call. It reads config, decides
 * whether to compress, and runs the pipeline:
 *
 *   1. Check mode (off → passthrough)
 *   2. Apply strategies (dedup, purge-errors)
 *   3. Check token budget (below threshold → passthrough)
 *   4. For each message: classify() → Router picks a Compressor → compress
 *   5. Aggregate stats and return.
 */

import { classify } from "./classifier.js";
import type { Compressor } from "./compressors/base.js";
import { Extract } from "./compressors/extract.js";
import { Verbatim } from "./compressors/verbatim.js";
import { Router } from "./router.js";
import type { CompressionStats } from "./stats.js";
import { makeStats } from "./stats.js";
import { DedupStrategy, PurgeErrorsStrategy } from "./strategies.js";
import type { Strategy } from "./strategies.js";
import { countMessageTokens } from "./tokens.js";
import { CONTENT_TYPE_VALUES, ContentType } from "./types.js";
import type { ChatMessage, LeanpromptConfig } from "./types.js";

export type { ChatMessage, LeanpromptConfig } from "./types.js";

const DEFAULT_THRESHOLD_TOKENS = 2000;

// Messages this recent are never handed to a lossy compressor: the tail of
// the conversation carries the live question/instructions, where dropping a
// sentence is costliest. Override via config.protect.lastTurns.
const DEFAULT_PROTECT_LAST_TURNS = 2;

type CompressorFactory = (config: LeanpromptConfig) => Compressor;

/**
 * Factories for Compressor names that appear in routing config. SelfLLM is
 * registered by the provider-client module (Phase C) via registerCompressor —
 * keeping the core middleware import-clean of any network code.
 */
const COMPRESSOR_FACTORIES = new Map<string, CompressorFactory>([
    ["verbatim", () => new Verbatim()],
    ["extract", (cfg) => new Extract({ ratio: cfg.extract?.ratio })],
]);

/** Register (or override) a compressor factory by routing name. */
export function registerCompressor(name: string, factory: CompressorFactory): void {
    COMPRESSOR_FACTORIES.set(name, factory);
}

/** Orchestrates compression across a single SDK call. */
export class Middleware {
    readonly config: LeanpromptConfig;
    private readonly active: boolean;
    private readonly threshold: number;
    private readonly protectLastTurns: number;
    private readonly router: Router;
    private readonly strategies: Strategy[];
    private readonly protector = new Verbatim();

    constructor(config: LeanpromptConfig = {}) {
        this.config = config;

        const mode = String(config.mode ?? "off").toLowerCase();
        this.active = !["off", "passthrough", "disabled"].includes(mode);

        this.threshold =
            config.trigger?.thresholdTokens ?? DEFAULT_THRESHOLD_TOKENS;
        this.protectLastTurns =
            config.protect?.lastTurns ?? DEFAULT_PROTECT_LAST_TURNS;

        this.router = this.buildRouter(config);
        this.strategies = this.active ? buildStrategies(config) : [];
    }

    /**
     * Protection rule: system messages and the last K turns are never handed
     * to a lossy compressor — they carry the live instructions/questions
     * where dropping a sentence is costliest.
     */
    private isProtected(index: number, total: number, msg: ChatMessage): boolean {
        if (msg.role === "system") {
            return true;
        }
        return index >= total - this.protectLastTurns;
    }

    compressMessages(messages: ChatMessage[]): [ChatMessage[], CompressionStats] {
        if (!this.active || messages.length === 0) {
            return [
                messages,
                makeStats({ method: messages.length === 0 ? "empty" : "passthrough" }),
            ];
        }

        for (const strategy of this.strategies) {
            messages = strategy.apply(messages);
        }
        if (messages.length === 0) {
            return [messages, makeStats({ method: "empty" })];
        }

        const inputTokens = countMessageTokens(messages);
        if (inputTokens < this.threshold) {
            return [
                messages,
                makeStats({
                    inputTokens,
                    outputTokens: inputTokens,
                    ratio: 1.0,
                    method: "below-threshold",
                }),
            ];
        }

        const outMsgs: ChatMessage[] = [];
        let totalIn = 0;
        let totalOut = 0;
        let totalCost = 0;
        const methods = new Set<string>();

        for (let i = 0; i < messages.length; i++) {
            const msg = messages[i]!;
            const compressor = this.isProtected(i, messages.length, msg)
                ? this.protector
                : this.router.route(classify(msg));
            const [compressed, stats] = compressor.compress([msg]);
            outMsgs.push(...compressed);
            totalIn += stats.inputTokens;
            totalOut += stats.outputTokens;
            totalCost += stats.costUsd;
            methods.add(stats.method);
        }

        return [outMsgs, aggregate(totalIn, totalOut, totalCost, methods)];
    }

    async compressMessagesAsync(
        messages: ChatMessage[],
    ): Promise<[ChatMessage[], CompressionStats]> {
        if (!this.active || messages.length === 0) {
            return [
                messages,
                makeStats({ method: messages.length === 0 ? "empty" : "passthrough" }),
            ];
        }

        for (const strategy of this.strategies) {
            messages = strategy.apply(messages);
        }
        if (messages.length === 0) {
            return [messages, makeStats({ method: "empty" })];
        }

        const inputTokens = countMessageTokens(messages);
        if (inputTokens < this.threshold) {
            return [
                messages,
                makeStats({
                    inputTokens,
                    outputTokens: inputTokens,
                    ratio: 1.0,
                    method: "below-threshold",
                }),
            ];
        }

        const outMsgs: ChatMessage[] = [];
        let totalIn = 0;
        let totalOut = 0;
        let totalCost = 0;
        const methods = new Set<string>();

        for (let i = 0; i < messages.length; i++) {
            const msg = messages[i]!;
            const compressor = this.isProtected(i, messages.length, msg)
                ? this.protector
                : this.router.route(classify(msg));
            const [compressed, stats] = await compressor.compressAsync([msg]);
            outMsgs.push(...compressed);
            totalIn += stats.inputTokens;
            totalOut += stats.outputTokens;
            totalCost += stats.costUsd;
            methods.add(stats.method);
        }

        return [outMsgs, aggregate(totalIn, totalOut, totalCost, methods)];
    }

    private buildRouter(config: LeanpromptConfig): Router {
        const router = new Router(null, new Verbatim());
        const routing = config.routing ?? {};

        for (const [ctypeStr, compressorName] of Object.entries(routing)) {
            if (!CONTENT_TYPE_VALUES.has(ctypeStr)) {
                console.warn(
                    `leanprompt: unknown content type ${JSON.stringify(ctypeStr)} in routing config; ignored`,
                );
                continue;
            }
            const factory = COMPRESSOR_FACTORIES.get(compressorName);
            if (factory === undefined) {
                console.warn(
                    `leanprompt: compressor ${JSON.stringify(compressorName)} not available; ` +
                        `falling back to default (verbatim) for ${ctypeStr}`,
                );
                continue;
            }
            router.register(ctypeStr as ContentType, factory(config));
        }

        // STRUCTURED always routes verbatim unless explicitly overridden:
        // sentence-segmenting JSON is meaningless, and even a gentle
        // extractive pass shreds it. (Deliberate deletion of the Python
        // reference's structured→gentler-ratio auto-route.)
        return router;
    }
}

function buildStrategies(config: LeanpromptConfig): Strategy[] {
    const strategiesCfg = config.strategies ?? {};
    const out: Strategy[] = [];

    if (strategiesCfg.dedup ?? true) {
        out.push(new DedupStrategy());
    }

    const purgeCfg = strategiesCfg.purgeErrors ?? true;
    if (purgeCfg) {
        let afterTurns = 4;
        if (typeof purgeCfg === "object") {
            afterTurns = purgeCfg.afterTurns ?? 4;
        }
        out.push(new PurgeErrorsStrategy(afterTurns));
    }

    return out;
}

function aggregate(
    totalIn: number,
    totalOut: number,
    totalCost: number,
    methods: Set<string>,
): CompressionStats {
    let method: string;
    if (methods.size === 0) {
        method = "empty";
    } else if (methods.size === 1) {
        method = methods.values().next().value as string;
    } else {
        method = "hybrid";
    }
    return makeStats({
        inputTokens: totalIn,
        outputTokens: totalOut,
        ratio: totalIn ? totalOut / totalIn : 1.0,
        method,
        costUsd: totalCost,
    });
}

File: rust/src/middleware.rs
//! Middleware — the compression orchestrator (parity-spec §9).

use std::collections::HashSet;

use crate::classifier::{classify, ContentType};
use crate::compressors::extract::Extract;
use crate::compressors::verbatim::Verbatim;
use crate::compressors::Compressor;
use crate::json::Value;
use crate::router::Router;
use crate::stats::CompressionStats;
use crate::strategies::{DedupStrategy, PurgeErrorsStrategy, Strategy};
use crate::tokens::count_message_tokens;

const DEFAULT_THRESHOLD_TOKENS: i64 = 2000;
const DEFAULT_PROTECT_LAST_TURNS: usize = 2;

/// Middleware configuration. Mirrors the reference config surface with
/// integer ratios (thousandths) instead of floats.
pub struct Config {
    /// "on"/"hybrid" activate; "off"/"passthrough"/"disabled" bypass.
    pub mode: String,
    pub threshold_tokens: i64,
    /// content-type name → compressor name ("verbatim" | "extract").
    pub routing: Vec<(String, String)>,
    pub extract_ratio_millis: i64,
    pub protect_last_turns: usize,
    pub dedup: bool,
    /// None disables purge; Some(n) purges errors older than n turns.
    pub purge_errors_after_turns: Option<usize>,
}

impl Default for Config {
    fn default() -> Self {
        Config {
            mode: "off".to_string(),
            threshold_tokens: DEFAULT_THRESHOLD_TOKENS,
            routing: Vec::new(),
            extract_ratio_millis: 500,
            protect_last_turns: DEFAULT_PROTECT_LAST_TURNS,
            dedup: true,
            purge_errors_after_turns: Some(4),
        }
    }
}

pub struct Middleware {
    active: bool,
    threshold: i64,
    protect_last_turns: usize,
    router: Router,
    protector: Verbatim,
    strategies: Vec<Box<dyn Strategy>>,
}

impl Middleware {
    pub fn new(config: Config) -> Self {
        let mode = config.mode.to_lowercase();
        let active = !matches!(mode.as_str(), "off" | "passthrough" | "disabled");

        let mut router = Router::new();
        for (ctype_str, compressor_name) in &config.routing {
            let Some(ctype) = ContentType::parse(ctype_str) else {
                eprintln!("leanprompt: unknown content type {ctype_str:?} in routing config; ignored");
                continue;
            };
            let compressor: Box<dyn Compressor> = match compressor_name.as_str() {
                "verbatim" => Box::new(Verbatim),
                "extract" => Box::new(Extract::new(config.extract_ratio_millis)),
                other => {
                    eprintln!(
                        "leanprompt: compressor {other:?} not available; falling back to default (verbatim) for {ctype_str}"
                    );
                    continue;
                }
            };
            router.register(ctype, compressor);
        }
        // STRUCTURED stays on the Verbatim default unless explicitly routed —
        // sentence-segmenting JSON is meaningless.

        let mut strategies: Vec<Box<dyn Strategy>> = Vec::new();
        if active {
            if config.dedup {
                strategies.push(Box::new(DedupStrategy));
            }
            if let Some(after_turns) = config.purge_errors_after_turns {
                strategies.push(Box::new(PurgeErrorsStrategy { after_turns }));
            }
        }

        Middleware {
            active,
            threshold: config.threshold_tokens,
            protect_last_turns: config.protect_last_turns,
            router,
            protector: Verbatim,
            strategies,
        }
    }

    pub fn compress_messages(&self, messages: &[Value]) -> (Vec<Value>, CompressionStats) {
        if !self.active || messages.is_empty() {
            let method = if messages.is_empty() { "empty" } else { "passthrough" };
            return (messages.to_vec(), CompressionStats::with_method(method));
        }

        let mut messages: Vec<Value> = messages.to_vec();
        for strategy in &self.strategies {
            messages = strategy.apply(messages);
        }
        if messages.is_empty() {
            return (messages, CompressionStats::with_method("empty"));
        }

        let input_tokens = count_message_tokens(&messages);
        if input_tokens < self.threshold {
            return (
                messages,
                CompressionStats::counted("below-threshold", input_tokens, input_tokens),
            );
        }

        let total = messages.len();
        let mut out = Vec::with_capacity(total);
        let mut total_in = 0;
        let mut total_out = 0;
        let mut total_cost = 0.0;
        let mut methods: HashSet<String> = HashSet::new();

        for (i, msg) in messages.iter().enumerate() {
            let compressor: &dyn Compressor = if self.is_protected(i, total, msg) {
                &self.protector
            } else {
                self.router.route(classify(msg))
            };
            let (compressed, stats) = compressor.compress(std::slice::from_ref(msg));
            out.extend(compressed);
            total_in += stats.input_tokens;
            total_out += stats.output_tokens;
            total_cost += stats.cost_usd;
            methods.insert(stats.method);
        }

        (out, aggregate(total_in, total_out, total_cost, &methods))
    }

    /// System messages and the last K turns are never handed to a lossy
    /// compressor — they carry the live instructions/questions.
    fn is_protected(&self, index: usize, total: usize, msg: &Value) -> bool {
        if msg.str_field("role") == "system" {
            return true;
        }
        index + self.protect_last_turns >= total
    }
}

fn aggregate(
    total_in: i64,
    total_out: i64,
    total_cost: f64,
    methods: &HashSet<String>,
) -> CompressionStats {
    let method = if methods.is_empty() {
        "empty".to_string()
    } else if methods.len() == 1 {
        methods.iter().next().unwrap().clone()
    } else {
        "hybrid".to_string()
    };
    let mut stats = CompressionStats::counted(&method, total_in, total_out);
    stats.cost_usd = total_cost;
    stats
}
